File: flutter-app/lib/canvas/drawing_controller.dart
import 'dart:ui';

import 'package:flutter/foundation.dart';

import 'stroke.dart';

/// 캔버스 상태(획 목록·도구·선택·undo/redo)를 들고 있는 컨트롤러.
class DrawingController extends ChangeNotifier {
  final DrawTool tool = DrawTool();

  List<Stroke> _strokes = [];
  final Set<int> _selection = {};
  final List<List<Stroke>> _undoStack = [];
  final List<List<Stroke>> _redoStack = [];

  List<Stroke> get strokes => _strokes;
  Set<int> get selection => _selection;
  bool get canUndo => _undoStack.isNotEmpty;
  bool get canRedo => _redoStack.isNotEmpty;

  /// 변경 직전에 호출해 현재 상태를 undo 스택에 쌓는다.
  void pushSnapshot() {
    _undoStack.add(List.of(_strokes));
    _redoStack.clear();
  }

  /// 이미 반영된 변경에 대해 시작 시점(base)을 undo 스택에 쌓는다(이동/크기조절).
  void commitFrom(List<Stroke> base) {
    _undoStack.add(base);
    _redoStack.clear();
    notifyListeners();
  }

  void setStrokes(List<Stroke> next) {
    _strokes = next;
    notifyListeners();
  }

  // ── 선택 ──────────────────────────────────────────

  void setSelection(Set<int> next) {
    _selection
      ..clear()
      ..addAll(next);
    notifyListeners();
  }

  void clearSelection() {
    if (_selection.isEmpty) return;
    _selection.clear();
    notifyListeners();
  }

  void deleteSelected() {
    if (_selection.isEmpty) return;
    pushSnapshot();
    _strokes = [
      for (var i = 0; i < _strokes.length; i++)
        if (!_selection.contains(i)) _strokes[i],
    ];
    _selection.clear();
    notifyListeners();
  }

  void recolorSelected(Color color) {
    if (_selection.isEmpty) return;
    pushSnapshot();
    _strokes = [
      for (var i = 0; i < _strokes.length; i++)
        _selection.contains(i)
            ? _strokes[i].copyWith(color: color)
            : _strokes[i],
    ];
    notifyListeners();
  }

  // ── undo / redo ───────────────────────────────────

  void undo() {
    if (_undoStack.isEmpty) return;
    _redoStack.add(_strokes);
    _strokes = _undoStack.removeLast();
    _selection.clear(); // 인덱스가 어긋날 수 있어 선택 해제
    notifyListeners();
  }

  void redo() {
    if (_redoStack.isEmpty) return;
    _undoStack.add(_strokes);
    _strokes = _redoStack.removeLast();
    _selection.clear();
    notifyListeners();
  }

  void clear() {
    if (_strokes.isEmpty) return;
    pushSnapshot();
    _strokes = [];
    _selection.clear();
    notifyListeners();
  }

  /// 도구 상태 변경 알림용(툴바에서 사용).
  void toolChanged() => notifyListeners();
}

File: ipad-app/DrawingController.swift
import Combine
import UIKit

/// 캔버스 상태(획 목록·도구·선택·undo/redo)를 들고 있는 컨트롤러.
final class DrawingController: ObservableObject {
    @Published var tool = DrawTool()
    @Published private(set) var strokes: [Stroke] = []
    @Published private(set) var selection: Set<Int> = []

    @Published private var undoStack: [[Stroke]] = []
    @Published private var redoStack: [[Stroke]] = []

    var canUndo: Bool { !undoStack.isEmpty }
    var canRedo: Bool { !redoStack.isEmpty }

    /// 변경 직전에 호출해 현재 상태를 undo 스택에 쌓는다.
    func pushSnapshot() {
        undoStack.append(strokes)
        redoStack.removeAll()
    }

    /// 이미 반영된 변경에 대해 시작 시점(base)을 undo 스택에 쌓는다(이동/크기조절).
    func commitFrom(_ base: [Stroke]) {
        undoStack.append(base)
        redoStack.removeAll()
    }

    func setStrokes(_ next: [Stroke]) {
        strokes = next
    }

    // MARK: - 선택

    func setSelection(_ next: Set<Int>) {
        selection = next
    }

    func clearSelection() {
        guard !selection.isEmpty else { return }
        selection = []
    }

    func deleteSelected() {
        guard !selection.isEmpty else { return }
        pushSnapshot()
        strokes = strokes.enumerated()
            .filter { !selection.contains($0.offset) }
            .map(\.element)
        selection = []
    }

    func recolorSelected(_ color: UIColor) {
        guard !selection.isEmpty else { return }
        pushSnapshot()
        for i in selection where strokes.indices.contains(i) {
            strokes[i].color = color
        }
    }

    // MARK: - undo / redo

    func undo() {
        guard let previous = undoStack.popLast() else { return }
        redoStack.append(strokes)
        strokes = previous
        selection = [] // 인덱스가 어긋날 수 있어 선택 해제
    }

    func redo() {
        guard let next = redoStack.popLast() else { return }
        undoStack.append(strokes)
        strokes = next
        selection = []
    }

    func clear() {
        guard !strokes.isEmpty else { return }
        pushSnapshot()
        strokes = []
        selection = []
    }
}
